File: src/dispatcher/provider.ts
import { NetworkConnection } from "./network-connection";
import { PrivateMetaInfo, PublicMetaInfo } from "./protocol/meta";
import { TaskResult } from "./protocol/task";
import { ProviderEstimator } from "./provider-estimator";
import { Task, TaskStatus } from "./task";
import { logger } from "./utils/logger";
import { NOOP } from "./utils/noop";

export const RETRY_ATTEMPTS = 3;
export const OFFLINE_TIMEOUT = 3000; // 3 seconds

export class Provider {
  public readonly estimator: ProviderEstimator;

  private online = true;

  private offlineTimeout: NodeJS.Timeout | undefined = undefined;

  private inProgress = new Set<Task>();

  constructor(
    public readonly id: string,
    private public_meta: PublicMetaInfo,
    private private_meta: PrivateMetaInfo,
    public readonly network_connection: NetworkConnection
  ) {
    this.estimator = new ProviderEstimator(public_meta, private_meta);

    network_connection.onMetaUpdated = (meta) => {
      this.updatePublicMeta(meta);
    };

    network_connection.onLostConnection = () => {
      this.startOffline();
    };

    network_connection.onConnectionRestored = () => {
      this.stopOffline();
    };

    network_connection.onTaskCompleted = (task, result) => {
      this.taskCompleted(task, result);
      this.onUpdated();
    };

    network_connection.onTaskFailed = (task, reason) => {
      this.taskFailed(task);
      this.onUpdated();
    };
  }

  private startOffline(): void {
    if (this.offlineTimeout) return;
    this.offlineTimeout = setTimeout(() => {
      this.dispose();
      this.onClosed();
    }, 1000);
    this.online = false;
    this.onUpdated();
  }

  private dispose(): void {
    this.network_connection.close();
    for (const task of this.inProgress) {
      task.setStatus(TaskStatus.FailedByProvider, { reason: "Provider is offline" });
      task.fail();
    }
    this.onClosed();
  }

  private stopOffline(): void {
    if (!this.offlineTimeout) return;
    clearTimeout(this.offlineTimeout);
    this.offlineTimeout = undefined;
    this.online = true;
    this.onUpdated();
  }

  updatePublicMeta(public_meta: PublicMetaInfo) {
    this.public_meta = public_meta;
    this.estimator.updatePublicMeta(public_meta);
    this.onUpdated();
  }

  updatePrivateMeta(private_meta: PrivateMetaInfo) {
    this.private_meta = private_meta;
    this.estimator.updatePrivateMeta(private_meta);
    this.onUpdated();
  }

  getQueueLength(): number {
    return this.inProgress.size;
  }

  scheduleTask(task: Task): void {
    task.setProviderId(this.id);
    console.log(`Task ${task.id} scheduled in provider ${this.id}`);
    this.estimator.addTask(task);
    this.asyncScheduleTask(task);
  }

  // TODO: Optimization batch sending
  private async asyncScheduleTask(task: Task): Promise<void> {
    for (let i = 0; i < RETRY_ATTEMPTS; i++) {
      try {
        await this.network_connection.sendTask(task);
        this.inProgress.add(task);
        return;
      } catch {
        task.setStatus(TaskStatus.SentFailed, { attempt: i });
      }
    }
    this.taskFailed(task, "Failed to send task");
  }

  abortTask(task: Task): void {
    this.taskFinished(task);

    (async () => {
      try {
        await this.network_connection.abortTask(task);
      } catch (e) {
        logger.error(e);
      }
    })();
  }

  taskFinished(task: Task): void {
    this.estimator.removeTask(task);
    this.inProgress.delete(task);
  }

  taskFailed(task: Task, reason?: string): void {
    if (!this.inProgress.has(task)) return;
    this.taskFinished(task);
    task.addFailedAttempt();
    task.setStatus(TaskStatus.FailedByProvider, { reason });
    task.fail();
  }

  taskCompleted(task: Task, result: TaskResult): void {
    if (!this.inProgress.has(task)) return;
    this.taskFinished(task);
    task.setStatus(TaskStatus.Completed);
    task.complete(result);
  }

  getMinCost(): number {
    if (!this.online) return Number.MAX_SAFE_INTEGER;
    return this.public_meta.min_cost;
  }

  onClosed: () => void = NOOP;

  onUpdated: () => void = NOOP;
}

File: backend/src/dispatcher/provider.py
from dispatcher.network_connection import NetworkConnection 
from dispatcher.protocol.meta import PrivateMetaInfo, PublicMetaInfo 
from dispatcher.protocol.task import TaskResult 
from dispatcher.provider_estimator import ProviderEstimator 
from dispatcher.task import Task, TaskStatus 
from dispatcher.utils.logger import logger 
from dispatcher.utils.noop import NOOP
from src.utils.timeouts import setTimeout, clearTimeout

RETRY_ATTEMPTS = 3 
OFFLINE_TIMEOUT = 3000 # 3 seconds

class Provider: 
    def init(self, id, public_meta: PublicMetaInfo, private_meta: PrivateMetaInfo, network_connection: NetworkConnection): 
        self.id = id 
        self.estimator = ProviderEstimator(public_meta, private_meta)

        self.online = True
        self.offline_timeout = None
        self.in_progress = set()

        self.public_meta = public_meta
        self.private_meta = private_meta
        self.network_connection = network_connection

        network_connection.on_meta_updated = self.update_public_meta
        network_connection.on_lost_connection = self.start_offline
        network_connection.on_connection_restored = self.stop_offline
        network_connection.on_task_completed = self.task_completed
        network_connection.on_task_failed = self.task_failed

def start_offline(self):
    if self.offline_timeout:
        return
    self.offline_timeout = setTimeout(self.dispose, 1000)
    self.online = False
    self.on_updated()

def dispose(self):
    self.network_connection.close()
    for task in self.in_progress:
        task.set_status(TaskStatus.FailedByProvider, {"reason": "Provider is offline"})
        task.fail()
    self.on_closed()

def stop_offline(self):
    if not self.offline_timeout:
        return
    clearTimeout(self.offline_timeout)
    self.offline_timeout = None
    self.online = True
    self.on_updated()

def update_public_meta(self, public_meta: PublicMetaInfo):
    self.public_meta = public_meta
    self.estimator.update_public_meta(public_meta)
    self.on_updated()

def update_private_meta(self, private_meta: PrivateMetaInfo):
    self.private_meta = private_meta
    self.estimator.update_private_meta(private_meta)
    self.on_updated()

def get_queue_length(self):
    return len(self.in_progress)

def schedule_task(self, task: Task):
    task.set_provider_id(self.id)
    print(f"Task {task.id} scheduled in provider {self.id}")
    self.estimator.add_task(task)
    self.async_schedule_task(task)

async def async_schedule_task(self, task: Task):
    for i in range(RETRY_ATTEMPTS):
        try:
            await self.network_connection.send_task(task)
            self.in_progress.add(task)
            return
        except:
            task.set_status(TaskStatus.SentFailed, {"attempt": i})
    self.task_failed(task, "Failed to send task")

def abort_task(self, task: Task):
    self.task_finished(task)

    async def abort():
        try:
            await self.network_connection.abort_task(task)
        except Exception as e:
            logger.error(e)
    
    abort()

def task_finished(self, task: Task):
    self.estimator.remove_task(task)
    self.in_progress.remove(task)

def task_failed(self, task: Task, reason = None):
    if task not in self.in_progress:
        return
    self.task_finished(task)
    task.add_failed_attempt()
    task.set_status(TaskStatus.FailedByProvider, {"reason": reason})
    task.fail()

def task_completed(self, task: Task, result: TaskResult):
    if task not in self.in_progress:
        return
    self.task_finished(task)
    task.set_status(TaskStatus.Completed)
    task.complete(result)

def get_min_cost(self):
    if not self.online:
        return float('inf')
    return self.public_meta.min_cost

def set_on_closed(self, callback):
    self.on_closed = callback

def set_on_updated(self, callback):
    self.on_updated = callback